cmd/srcd/daemon/daemon.go

```diff
@@ -141,3 +163,112 @@ func createDaemon(workdir, datadir string) docker.StartFunc {
 		return docker.Start(ctx, config, host, daemonName)
 	}
 }
+
+// GetCompatibleTag returns semver compatible tag of daemon image by cli version
+// and boolean flag if there are any newer versions with breaking changes
+func GetCompatibleTag(cliVersion string) (string, bool, error) {
+	if cliVersion == "" || cliVersion == "dev" {
+		return "latest", false, nil
+	}
+
+	cliV, err := semver.ParseTolerant(cliVersion)
+	if err != nil {
+		return "", false, err
+	}
+
+	tags, err := getTags()
+	if err != nil {
+		return "", false, err
+	}
+
+	var breakingV semver.Version
+	if cliV.Major >= 1 {
+		breakingV = semver.Version{Major: cliV.Major + 1}
+	} else {
+		breakingV = semver.Version{Minor: cliV.Minor + 1}
+	}
+
+	var newestV semver.Version
+	var hasNewBreakingTag bool
+	for _, tag := range tags {
+		v, err := semver.ParseTolerant(tag)
+		if err != nil {
+			continue
+		}
+
+		// skip pre-releases
+		if len(v.Pre) > 0 {
+			continue
+		}
+
+		// skip old versions
+		if v.LT(cliV) {
+			continue
+		}
+
+		// skip anything that breaks
+		if v.GTE(breakingV) {
+			hasNewBreakingTag = true
+			continue
+		}
+
+		if v.GT(newestV) {
+			newestV = v
+		}
+	}
+
+	if newestV.Equals(semver.Version{}) {
+		return "", false, errors.New("can't find compatible daemon image")
+	}
+
+	return "v" + newestV.String(), hasNewBreakingTag, nil
+}
+
+func getTags() ([]string, error) {
+	c := &http.Client{}
+
+	v := url.Values{
+		"service": []string{"registry.docker.io"},
+		"scope":   []string{fmt.Sprintf("repository:%s:pull", daemonImage)},
+	}
+	r, err := c.Get(fmt.Sprintf("https://auth.docker.io/token?%s", v.Encode()))
+	if err != nil {
+		return nil, errors.Wrap(err, "can't authorize in docker registry")
+	}
+
+	if r.StatusCode != http.StatusOK {
+		return nil, errors.New("incorrect status code")
+	}
+
+	var authResp struct {
+		Token string
+	}
+	jd := json.NewDecoder(r.Body)
+	err = jd.Decode(&authResp)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't parse authorization response from docker registry")
+	}
+
+	req, _ := http.NewRequest("GET", fmt.Sprintf("https://registry-1.docker.io/v2/%s/tags/list", daemonImage), nil)
+	req.Header.Add("Authorization", "Bearer "+authResp.Token)
+
+	r, err = c.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't request list of tags in docker registry")
+	}
+
+	if r.StatusCode != http.StatusOK {
+		return nil, errors.New("incorrect status code")
```

Comment:
This error could have more context, like "unexpected status code %d while requesting the list of tags in docker registry"